rust/kernel/virtio.rs

```diff
@@ -0,0 +1,187 @@
+            // `vdev`.
+            let mut dev = unsafe { Device::from_ptr(vdev) };
+            let data = T::probe(&mut dev)?;
+            // SAFETY: `vdev` is guaranteed to be a valid, non-null pointer.
+            unsafe{(*vdev).priv_ = T::Data::into_pointer(data) as _;}
+            Ok(0)
+        }
+    }
+
+    extern "C" fn remove_callback(vdev: *mut bindings::virtio_device) {
+        // SAFETY: `vdev` is guaranteed to be a valid, non-null pointer.
+        let ptr = unsafe { *vdev }.priv_;
+        // SAFETY:
+        //   - we allocated this pointer using `T::Data::into_pointer`,
+        //     so it is safe to turn back into a `T::Data`.
+        //   - the allocation happened in `probe`, no-one freed the memory,
+        //     `remove` is the canonical kernel location to free driver data. so OK
+        //     to convert the pointer back to a Rust structure here.
+        let data = unsafe { T::Data::from_pointer(ptr) };
+        T::remove(&data);
+        <T::Data as driver::DeviceRemoval>::device_remove(&data);
+    }
+}
+
+/// A virtio driver.
+pub trait Driver {
+    /// Data stored on device by driver.
+    type Data: PointerWrapper + Send + Sync + driver::DeviceRemoval = ();
+
+    /// The table of device ids supported by the driver.
+    const ID_TABLE: Option<driver::IdTable<'static, DeviceId, ()>> = None;
+
+    /// Probes for the device with the given id.
+    fn probe(dev: &mut Device) -> Result<Self::Data>;
+
+    /// Cleans any resources up that are associated with the device.
+    ///
+    /// This is called when the driver is detached from the device.
+    fn remove(_data: &Self::Data) {}
+}
+
+/// A Virtio device.
+///
+/// # Invariants
+///
+/// The field `ptr` is non-null and valid for the lifetime of the object.
+pub struct Device {
+    ptr: *mut bindings::virtio_device,
+}
+
+impl Device {
+    /// Creates a new device from the given pointer.
+    ///
+    /// # Safety
+    ///
+    /// `ptr` must be non-null and valid. It must remain valid for the lifetime of the returned
+    /// instance.
+    unsafe fn from_ptr(ptr: *mut bindings::virtio_device) -> Self {
+        // INVARIANT: The safety requirements of the function ensure the lifetime invariant.
+        Self { ptr }
+    }
+}
+
+// SAFETY: The device returned by `raw_device` is the raw virtio device.
+unsafe impl device::RawDevice for Device {
+    fn raw_device(&self) -> *mut bindings::device {
+        // SAFETY: By the type invariants, we know that `self.ptr` is non-null and valid.
+        unsafe { &mut (*self.ptr).dev }
+    }
+}
+
+/// Declares a kernel module that exposes a single virtio driver.
+///
+/// # Examples
+///
+/// ```ignore
+/// # use kernel::{virtio, define_of_id_table, module_virtio_driver};
+/// #
+/// struct MyDriver;
+/// impl virtio::Driver for MyDriver {
+///     // [...]
+/// #   fn probe(_dev: &mut virtio::Device) -> Result {
+/// #       Ok(())
+/// #   }
+/// #   define_of_id_table! {(), [
```

Comment:
This does not work. I misunderstand the usage of define_of_id_table. I write a new macro `define_virtio_id_table` in the commit.

Reply:
BTW, Is there a easy way to compile the example code in the comment except by writing the example code in the new driver file?

Reply:
Making the test non-`ignore` would be the ideal, wherever possible. Otherwise, if you provide a `samples/rust/` example, you can have the complete code there, have it tested, and then it is simpler to copy-paste that working code into the docs here knowing it actually compiles.

Reply:
> Making the test non-`ignore` would be the ideal, wherever possible. Otherwise, if you provide a `samples/rust/` example, you can have the complete code there, have it tested, and then it is simpler to copy-paste that working code into the docs here knowing it actually compiles.

Thanks! I will try to make the comment code non-`ignore`.